File: CSharpCodeBase/components/thing/treebehaviorcomponent.cs
using System;
 
 using System.Collections.Generic;
 
 using System.Linq;
 
 using System.Text;
 
 using UnityEngine;
 
 using System.Collections;
 
 using MainGame.core;
 
 using UnityEngine.EventSystems;
 
 namespace MainGame{
 public class treebehaviorcomponent :component{
 public void init(self, entity);
       Component.init(self, entity);
       this.entity.treeBehavior = self;
       this.animator = this.entity.mecanim;
       this.deathTimeout = this.entity.config:Get("deathTimeout");
       this.fallenTreeTransformName = this.entity.config:Get("rigidbodyTransform");
       this.stumpTransformName = this.entity.config:Get("stumpTransform");
       
       this.stumpTransform = this.entity.gameObject.transform:Find(this.stumpTransformName);
       this.fallenTreeTransform = this.entity.gameObject.transform:Find(this.fallenTreeTransformName).gameObject.transform;
 }})
 public void OnDeath(damageData){
   if(this.entity.stateManager.stateName[this.entity.stateManager.currentState] == "fallenTree"   ){ 
     self:SaveTransform();
     Transform.SetColliderState(this.stumpTransform, true);
     RigidbodyUtils.MoveDown(this.fallenTreeTransform.gameObject, damageData.source.gameObject);
   }
 }
 public void SaveTransform(){
   this.localPosition = this.fallenTreeTransform.localPosition;
   this.localRotation = this.fallenTreeTransform.localRotation;
 }
 public void RestoreTransform(){
   if(this.localPosition  &&  this.localRotation  ){ 
     this.fallenTreeTransform.localPosition = this.localPosition;
     this.fallenTreeTransform.localRotation = this.localRotation;
   }
 }
 public void AfterFall(result){
   this.entity.stateManager:SetStateByNumber(#this.entity.stateManager.stateName);
   self:RestoreTransform();
   this.entity:Destroy();
   result:Complete();
 }
 public void OnFragileForce(){
     Transform.SetColliderState(this.stumpTransform, false);
 }
 }}

File: CSharpCodeBase/components/thing/treebehaviorcomponent.lua
local Class = require 'utils.hump.class'
local Component = require 'components.component'

local TreeBehaviorComponent = Class({__includes=Component, init=function(self, entity)
      Component.init(self, entity)
      self.entity.treeBehavior = self
      self.animator = self.entity.mecanim
      self.deathTimeout = self.entity.config:Get('deathTimeout')
      self.fallenTreeTransformName = self.entity.config:Get('rigidbodyTransform')
      self.stumpTransformName = self.entity.config:Get('stumpTransform')
      
      self.stumpTransform = self.entity.gameObject.transform:Find(self.stumpTransformName)
      self.fallenTreeTransform = self.entity.gameObject.transform:Find(self.fallenTreeTransformName).gameObject.transform
end})

function TreeBehaviorComponent:OnDeath(damageData)
  if self.entity.stateManager.stateName[self.entity.stateManager.currentState] == 'fallenTree'  then
    self:SaveTransform()
    Transform.SetColliderState(self.stumpTransform, true)
    RigidbodyUtils.MoveDown(self.fallenTreeTransform.gameObject, damageData.source.gameObject)
  end
end

function TreeBehaviorComponent:SaveTransform()
  self.localPosition = self.fallenTreeTransform.localPosition
  self.localRotation = self.fallenTreeTransform.localRotation
end

function TreeBehaviorComponent:RestoreTransform()
  if self.localPosition and self.localRotation then
    self.fallenTreeTransform.localPosition = self.localPosition
    self.fallenTreeTransform.localRotation = self.localRotation
  end
end

function TreeBehaviorComponent:AfterFall(result)
  self.entity.stateManager:SetStateByNumber(#self.entity.stateManager.stateName)
  self:RestoreTransform()
  self.entity:Destroy()
  result:Complete()
end

function TreeBehaviorComponent:OnFragileForce()
    Transform.SetColliderState(self.stumpTransform, false)
end

return TreeBehaviorComponent